Change Raw now works with Enter.

codedmart opened the pull request:
Reviews welcome, this should resolve #157.

alpmestan commented:
LGTM! I definitely think that despite the code breakage that this is going to introduce, it will overall make "the servant experience" less surprising. This `Enter` instance is probably worth it.

@jkarni Feel free to merge if/when you're happy with this PR.

codedmart commented:
@jkarni I can rebase/fix conflicts Monday morning. Any feedback or suggestions on this one?

jkarni commented on servant-server/src/Servant/Utils/StaticFiles.hs:
```diff
@@ -35,10 +36,14 @@ import Filesystem.Path.CurrentOS (decodeString)
+serveDirectory = serveDirectoryWith . defaultFileServerSettings .
 #if MIN_VERSION_wai_app_static(3,1,0)
-    staticApp . defaultFileServerSettings . addTrailingPathSeparator
+  id .
```
Not that is matters much, but couldn't you use `#if !MIN... decodeString .` and not have an `else` clause (and no `id .`)?

codedmart replied:
Yes good point. Thanks!

jkarni commented on servant/src/Servant/API/Raw.hs:
```diff
@@ -11,4 +12,5 @@ import Data.Typeable (Typeable)
 -- this can also be used with 'Servant.Utils.StaticFiles.serveDirectory' to serve
 -- static files stored in a particular directory on your filesystem
-data Raw deriving Typeable
+newtype Raw a (m :: * -> *) = Raw a
```
I'd write out all the instances `Identity` (from [Data.Functor.Identity](https://hackage.haskell.org/package/base-4.8.0.0/docs/Data-Functor-Identity.html)) here. And maybe also have an `unRaw` record.

I think the documentation should say something like 

> the `m` type parameter is a phantom type used to allow <http://hackage.haskell.org/package/servant-server/docs/Servant-Server.html#g:4 `enter`> to treat endpoints uniformly. 

We're going to need a better explanation for that, but I haven't really figured out how best to phrase it...

jkarni replied:
[Tagged](http://hackage.haskell.org/package/tagged-0.7.2/docs/Data-Tagged.html) may have a few more instances that we should also add.

jkarni commented on servant/test/Servant/Utils/LinksSpec.hs:
```diff
@@ -26,7 +26,7 @@ type TestApi =
   :<|> "post" :> ReqBody '[JSON] 'True :> Post '[JSON] ()
   :<|> "delete" :> Header "ponies" String :> Delete '[JSON] ()
-  :<|> "raw" :> Raw
+  :<|> "raw" :> Raw () IO
```
Can all these `IO`s be just `m`? Semantically, I would expect it to be `EitherT ServantErr IO`, but I guess when `enter` isn't used it doesn't matter?

jkarni commented:
I mentioned this on IRC, but having `Raw` take an explicit param gives us the ability to have documentation that's more tailored to the specific details (e.g. that it servers a specific directory, or that it's a websocket endpoint). I think maybe that should be a different PR, but if so, we should open an issue so we don't forget.

codedmart commented:
@jkarni @alpmestan I added `unRaw` and removed the redundant id in `StaticFiles.hs`. If you do come up with a solution to swap `Raw` I am happy to finish this up, but I will be out for the remainder of this week and can't get to it till next week. Feel free to make the changes if you need it done sooner. Thanks!

codedmart commented:
I rebased master into here as well so it is up to date.

codedmart commented:
@jkarni I thought I had left this here before, but http://lpaste.net/138722#line27. I am not sure what is going on, but it is only a problem with ghc 7.8. Any thoughts/ideas?

jkarni commented on servant-server/src/Servant/Server/Internal/Enter.hs:
```diff
@@ -33,13 +36,15 @@ import           Servant.API
 class Enter typ arg ret | typ arg -> ret, typ ret -> arg where
     enter :: arg -> typ -> ret
 
+data IsRaw
```
This is no longer necessary, I think

codedmart replied:
Yes I will remove it. Thanks!

jkarni commented:
@alpmestan @kosmikus Any thoughts on the error codedmart posted above? It seems 7.8 is picking the *more* general instance, which is baffling to me.

alpmestan commented:
No clue to what's going on here. What if we switch the `OVERLAPPABLE`/`OVERLAPPING`? I.e `OVERLAPPABLE` for the "general case" and `OVERLAPPING` for the `Raw`s?

codedmart commented:
@alpmestan That won't matter because `OVERLAPPABLE/OVERLAPPING` only apply to 7.10.

alpmestan commented:
Oh right the problem is with 7.8, sorry.

codedmart commented:
@jkarni @alpmestan Don't merge yet if the tests pass. As per my convo with @jkarni I am going to add a few more complex api test to check Enter changes.

jkarni commented:
Closing in favour of #248
